text/0000-let-else.md

```diff
@@ -0,0 +1,520 @@
+    return Err(format_err_status!(
+        422,
+        "GeoJSON was not a Feature Collection",
+    ));
+};
+```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A common pattern in non-trivial code where static guarantees can not be fully met (e.g. I/O, network or otherwise) is to check error cases when possible before proceeding,
+and "return early", by constructing an error `Result` or an empty `Option`, and returning it before the "happy path" code.
+
+This pattern serves no practical purpose to a computer, but it is helpful for humans interacting with the code.
+Returning early helps improve code clarity in two ways:
+- Ensuring the returned result in near the conditional, visually, as the following logic may be lengthy.
+- Reduces rightward shift, as the error return is now in the block, rather than the following logic.
+
+This RFC proposes _(Rust provides)_ an extension to `let` assignment statements to help with this pattern, an `else { }` which can follow a pattern match
+as a `let` assigning statement:
+
+```rust
+let Some(a) = an_option else {
+    // Called if `an_option` is not `Option::Some(T)`.
+    // This block must diverge (stop executing the existing context to the parent block or function).
+    return;
+};
+
+// `a` is now in scope and is the type which the `Option` contained.
+```
+
+This is a counterpart to `if let` expressions, and the pattern matching works identically, except that the value from the pattern match
+is assigned to the surrounding scope rather than the block's scope.
+
+# Reference-level explanations
+[reference-level-explanation]: #reference-level-explanation
+
+let-else is syntactical sugar for `match` where the non-matched case diverges.
+```rust
+let pattern = expr else {
+    /* diverging expr */
+};
+```
+desugars to
+```rust
+let (each, binding) = match expr { 
+    pattern => (each, binding),
+    _ => { 
+        /* diverging expr */
+    }
+};
+```
+
+Any expression may be put into the expression position except an `if {} else {}` as explain below in [drawbacks][].
+While `if {} else {}` is technically feasible this RFC proposes it be disallowed for programmer clarity to avoid an `... else {} else {}` situation.
+Rust already provides us with such a restriction, [`ExpressionWithoutBlock`][expressions].
+
+Any pattern that could be put into if-let's pattern position can be put into let-else's pattern position, except a match to a boolean when
+a lazy boolean operator is present (`&&` or `||`), for reasons noted in [future-possibilities][].
+
+If the pattern is irrefutable, rustc will emit the `irrefutable_let_patterns` warning lint, as it does with an irrefutable pattern in an `if let`.
+
+The `else` block must diverge. This could be a keyword which diverges (returns `!`), or a panic.
+This likely necessitates a new subtype of `BlockExpression`, something like `BlockExpressionDiverging`.
+Allowed keywords:
+- `return`
+- `break`
+- `continue`
```

Comment:
I'm not sure what you're getting at with "allowed keywords" here.  If the restriction is type-based, not syntactic, there's no need to mention keywords or panics or similar.

I think you either want this reference-level explanation to
- define a typing rule for the `let-else` construct, or
- define the desugaring in such a way that it include the type restriction.

For example, rather than just `_ => { /* diverging expr */ }` in the earlier desugaring, have it be something like
```rust
_ => {
    enum Nope {}
    let nope: Nope = { /* diverging expr */ };
    match nope {}
}
```

Reply:
It can't use `Nope` because the type is observable. It could use `!`, although the desugaring would not be allowed on current stable:
```rust
_ => {
    let nope: ! = { /* diverging expr */ };
    match nope {}
}
```
You can get around this with some type inference hacks:
```rust
_ => {
    let nope = { /* diverging expr */ };
    if false {
        fn f<T>(t: &T, f: fn() -> T) {}
        fn g() -> ! { loop {} }
        f(&nope, g)
    }
    match nope {}
}
```

Reply:
It should be fine for desugaring to use unstable constructs, right?